Multiple devices support for opentherm_gw

mvn23 opened the pull request:
## Breaking Change:
<!-- What is breaking and why we have to break it. Remove this section only if it was NOT a breaking change. -->
- Breaks configuration structure: we need to specify and identify more than one gateway.
- Breaks configuration structure: `monitored_variables` option has been removed, all sensors are added by default.
- Breaks services: add a required parameter to all services to identify the gateway upon which the service will be called.
- Breaks entity_ids: new format to keep entity_ids unique for each configured gateway.

## Description:
Add support for more than one OpenTherm Gateway device.

**Related issue (if applicable):** N/A

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#9176

## Example entry for `configuration.yaml` (if applicable):
```yaml
opentherm_gw:
  living_room:
    device: /dev/ttyUSB0
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly ([example][ex-manifest]).
  - [x] New dependencies have been added to `requirements` in the manifest ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
 N/A

[ex-manifest]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mobile_app/manifest.json
[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mobile_app/manifest.json#L5
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23
[manifest-docs]: https://developers.home-assistant.io/docs/en/development_checklist.html#_the-manifest-file_

mvn23 committed:
Breaking change: Rewrite opentherm_gw to add support for more than one OpenTherm Gateway.
Breaks config layout and child entity ids and adds a required parameter to all service calls (gateway_id).

mvn23 committed:
Add schema and parameter description for service opentherm_gw.reset_gateway.

mvn23 committed:
Add optional name attribute in config to be used for friendly names.
Fix bugs in binary_sensor and climate platforms.

mvn23 committed:
pylint fixes

mvn23 committed:
Remove unused variables.

mvn23 committed:
Update manifest.json, remove REQUIREMENTS from .py file

mvn23 committed:
Update CODEOWNERS

mvn23 commented:
Updated to be compatible with #23810

mvn23 commented:
Thanks for the review, I will address the issues over the coming days.

mvn23 commented on homeassistant/components/opentherm_gw/__init__.py:
```diff
@@ -19,340 +20,375 @@
+async def register_services(hass):
     """Register services for the component."""
-    gw_vars = hass.data[DATA_OPENTHERM_GW][DATA_GW_VARS]
+    import pyotgw.vars as gw_vars
```
I thought dependency imports were only allowed inside functions that use them? Or does that rule expand to modules as well?

MartinHjelmare replied:
This changed after we introduced manifest.json. We can now import 3rd party modules and packages at the top of the module. This can make some more complicated structures, that were required before, obsolete.

mvn23 committed:
Address issues that were brought up (requested changes):
- Move imports to module level
- Change certain functions from async to sync
- Move constants to const.py (new file)
- Call gateway setup from outside of __init__()
- Move validation of monitored_variables to config schema

mvn23 commented:
Updated and rebased, all mentioned issues should be addressed.

MartinHjelmare commented on homeassistant/components/opentherm_gw/__init__.py:
```diff
@@ -2,9 +2,18 @@
 
 from homeassistant.components.binary_sensor import DOMAIN as COMP_BINARY_SENSOR
+from homeassistant.components.climate import DOMAIN as COMP_CLIMATE
+from homeassistant.components.opentherm_gw.const import (
```
Please make all intra package imports relative. This will allow custom integrations to override builtin integrations.

```py
from .const import (...)
```

mvn23 replied:
I used relative imports while it was a WIP, decided to change to the absolute form to make the imports more uniform. I will change it back with the next commit.

MartinHjelmare commented on homeassistant/components/opentherm_gw/binary_sensor.py:
```diff
@@ -3,116 +3,51 @@
                                  self.receive_report)
 
-    async def receive_report(self, status):
+    def receive_report(self, status):
```
This must be decorated with `@callback` imported from core.py to make sure this callback is run in the event loop.

mvn23 replied:
Fixed in 83ea38d, and applied in the other modules as well (sensor.py, climate.py).

MartinHjelmare commented on homeassistant/components/opentherm_gw/__init__.py:
```diff
@@ -15,344 +24,251 @@
         vol.Optional(CONF_CLIMATE, default={}): CLIMATE_SCHEMA,
-        vol.Optional(CONF_MONITORED_VARIABLES, default=[]): vol.All(
-            cv.ensure_list, [cv.string]),
+        vol.Optional(CONF_MONITORED_VARIABLES, default=[]): [vol.Any(
```
There was a recent architecture decision that we should avoid user selectable monitored variables: https://github.com/home-assistant/architecture/blob/master/adr/0003-monitor-condition-and-data-selectors.md.

Can we motivate keeping this option here after reading the decision or should we remove the option?

mvn23 replied:
For the last few weeks/months, I haven't really kept up with HA development so I was not aware of this. All data is obtained from the device regardless of the configured `monitored_variables`, so the option is just there to reduce the amount of entities the integration exposes. As lovelace is now the preferred way to 'filter' this information, I will remove the option to select which data is exposed and create all entities by default.
home-assistant/home-assistant.io#9176 will be updated accordingly as well.

mvn23 committed:
Address requested changes:
- Make module imports relative
- Move more functions from async to sync, decorate with @callback where necessary
- Remove monitored_variables option, add all sensors by default

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare commented:
Please update the breaking change paragraph in the PR description, regarding removing monitored variables.

Ping me when that is done and build passes and I can merge.

mvn23 commented:
This should be good to go. Thanks for the help @MartinHjelmare
